### ISSUE TITLE
brod_group_subscriber_v2 doesn't always stop consumers for partitions on termination

### ISSUE THREAD

**Issue by epsylonix** [kafka4beam/brod#433]
Hello!

I've encountered a problem when multiple instances of a service started consuming data from the same partitions.

As far as I can understand, what happened was `brod_group_coordinator` crashed and that in turn crashed `brod_group_subscriber_v2`.  Since `brod_group_subscriber_v2` `unlink`s its workers, the workers continued to consume data using old assignments. I'm not sure if that would work given that they were started with a different generation, but here is what I could get from logs and from remote connection to the service:

new assignments were received, `brod_group_subscriber_v2` started workers for every partition, but then `brod_group_coordinator` crashed 
```
2021-02-17 01:47:54.161 [info] pid=<0.16781.4032>  Group member (some-group,coor=#PID<0.16781.4032>,cb=#PID<0.16940.4032>,generation=22):
assignments received:
  some-topic:
    partition=0 begin_offset=109503
    partition=4 begin_offset=20271
    partition=8 begin_offset=35433
2021-02-17 01:47:54.162 [info] pid=<0.22956.4032>  Starting group_subscriber_worker: %{commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>, group_id: "some-group", partition: 0, topic: "some-topic"}
Offset: 109503
Pid: #PID<0.22956.4032>

2021-02-17 01:47:54.162 [info] pid=<0.16781.4032>  Group member (some-group,coor=#PID<0.16781.4032>,cb=#PID<0.16940.4032>,generation=22):
Leaving group, reason: :hb_timeout

2021-02-17 01:47:54.162 [info] pid=<0.24237.4032>  Starting group_subscriber_worker: %{commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>, group_id: "some-group", partition: 4, topic: "some-topic"}
Offset: 20271
Pid: #PID<0.24237.4032>

2021-02-17 01:47:54.163 [info] pid=<0.24692.4032>  Starting group_subscriber_worker: %{commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>, group_id: "some-group", partition: 8, topic: "some-topic"}
Offset: 35433
Pid: #PID<0.24692.4032>

2021-02-17 01:47:54.163 [info] pid=<0.2230.0>  Fetch error some-topic-8: :not_leader_for_partition
** (stop) :hb_timeout
Last message: :lo_cmd_send_heartbeat
```

then everything got restarted and new assignments for different partitions were received, but workers for old partitions were not stopped:

```
2021-02-17 01:47:59.166 [info] pid=<0.22810.4032>  Group member (some-group,coor=#PID<0.22810.4032>,cb=#PID<0.24452.4032>,generation=23):
elected=false
2021-02-17 01:47:59.169 [info] pid=<0.22810.4032>  Group member (some-group,coor=#PID<0.22810.4032>,cb=#PID<0.24452.4032>,generation=23):
assignments received:
  some-topic:
    partition=3 begin_offset=17157
    partition=7 begin_offset=27245
    partition=11 begin_offset=31523
2021-02-17 01:47:59.170 [info] pid=<0.24780.4032>  Starting group_subscriber_worker: %{commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>, group_id: "some-group", partition: 3, topic: "some-topic"}
Offset: 17157
Pid: #PID<0.24780.4032>

2021-02-17 01:47:59.171 [info] pid=<0.23822.4032>  Starting group_subscriber_worker: %{commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>, group_id: "some-group", partition: 7, topic: "some-topic"}
Offset: 27245
Pid: #PID<0.23822.4032>

2021-02-17 01:47:59.172 [info] pid=<0.24652.4032>  Starting group_subscriber_worker: %{commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>, group_id: "some-group", partition: 11, topic: "some-topic"}
Offset: 31523
Pid: #PID<0.24652.4032>
```

for example I could see a live topic subscriber for partition 8 that was not supposed to be active:
```
Process.info(:erlang.list_to_pid('<0.24692.4032>'))            
[
  current_function: {:gen_server, :loop, 7},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 0,
  links: [],
  dictionary: [
    "$initial_call": {:brod_topic_subscriber, :init, 1},
    "$ancestors": [#PID<0.16940.4032>, #PID<0.2208.0>, App.Supervisor,
     #PID<0.1959.0>]
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.1958.0>,
  total_heap_size: 13544,
  heap_size: 6772,
  stack_size: 10,
  reductions: 210446362,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 43
  ],
  suspending: []
]

{:state, :"kafka_consumer_client_some-group",
 #Reference<0.3387090685.283901953.155268>, some-topic",
 [
   {:consumer, 8, #PID<0.2230.0>, #Reference<0.3387090684.3205758978.12960>,
    26599544, 26599544}
 ], :brod_group_subscriber_worker,
 {:state,
  %{
    begin_offset: 26575433,
    cb_config: nil,
    cb_module: App.Consumer,
    commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>,
    group_id: "some-group",
    partition: 8,
    topic: some-topic"
  }, App.Consumer, nil,
  #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>},
 :message_set}
```
the first ancestor of this process was `<0.16940.4032>`, the `brod_group_subscriber_v2`, and it was not running at the moment:
```
:erlang.is_process_alive(:erlang.list_to_pid('<0.16940.4032>'))
false
```

**Comment by epsylonix** [comment]
at the same time another instance stopped committing offsets while I can see in logs that all messages are being processed. I'm not sure if this is related or not. Here is a state of `brod_topic_subscriber` for one of the partitions:

```
:sys.get_state(:erlang.list_to_pid('<0.15748.4031>'))
{:state, :"kafka_consumer_client_some-group",
 #Reference<0.3846263541.663224321.126968>, 
"mf-vpp-eh-contacts",
 [
   {:consumer, 3,
    {:down, '2021-02-17:13:20:41.751395',
     {:already_subscribed_by, #PID<0.10137.4031>}}, :undefined, 27189,
    :undefined}
 ], :brod_group_subscriber_worker,
 {:state,
  %{
    begin_offset: 27190,
    cb_config: nil,
    cb_module: App.Consumer,
    commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>,
    group_id: "some-group",
    partition: 3,
    topic: "some-topic"
  }, App.Consumer, nil,
  #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>},
 :message_set}
```

It only commited to offset 27190, and 27189 is current offset, all messages from 27190 to 27189 were processed successfully, callback module returns `{:ok, :commit, state}`

**Comment by epsylonix** [comment]
at the same time another instance stopped committing offsets while I can see in logs that all messages are being processed. I'm not sure if this is related or not. Here is a state of `brod_topic_subscriber` for one of the partitions:

```
:sys.get_state(:erlang.list_to_pid('<0.15748.4031>'))
{:state, :"kafka_consumer_client_some-group",
 #Reference<0.3846263541.663224321.126968>, 
"some-topic",
 [
   {:consumer, 3,
    {:down, '2021-02-17:13:20:41.751395',
     {:already_subscribed_by, #PID<0.10137.4031>}}, :undefined, 27189,
    :undefined}
 ], :brod_group_subscriber_worker,
 {:state,
  %{
    begin_offset: 27190,
    cb_config: nil,
    cb_module: App.Consumer,
    commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>,
    group_id: "some-group",
    partition: 3,
    topic: "some-topic"
  }, App.Consumer, nil,
  #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>},
 :message_set}
```

It only commited to offset 27190, and 27189 is NOT current offset

**Comment by epsylonix** [comment]
it seems to be related indeed, because `<0.10137.4031>` from `{:already_subscribed_by, #PID<0.10137.4031>}` - is another topic subscriber for the same partition, but somehow messages are processed in a different topic subscriber.

```
 Process.info(:erlang.list_to_pid('<0.10137.4031>'))

[
  current_function: {:gen_server, :loop, 7},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 0,
  links: [],
  dictionary: [
    "$initial_call": {:brod_topic_subscriber, :init, 1},
    "$ancestors": [#PID<0.2211.0>, #PID<0.2208.0>, App.Supervisor,
     #PID<0.1959.0>]
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.1958.0>,
  total_heap_size: 10957,
  heap_size: 4185,
  stack_size: 10,
  reductions: 142808819,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 78
  ],
  suspending: []
]

sys.get_state(:erlang.list_to_pid('<0.10137.4031>'))

{:state, :"kafka_consumer_client_some-group",
 #Reference<0.3846263541.660340737.50838>, "some-topic",
 [
   {:consumer, 3, #PID<0.2225.0>, #Reference<0.3846263541.660340737.50855>,
    26641056, 26641056}
 ], :brod_group_subscriber_worker,
 {:state,
  %{
    begin_offset: 26616766,
    cb_config: nil,
    cb_module: App.Consumer,
    commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>,
    group_id: "some-group",
    partition: 3,
    topic: "some-topic"
  }, App.Consumer, nil,
  #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>},
 :message_set}
```

**Comment by epsylonix** [comment]
it seems to be related indeed, because `<0.10137.4031>` from `{:already_subscribed_by, #PID<0.10137.4031>}` - is another topic subscriber for the same partition, and new messages are not processed there, but are not commited for some reason

```
 Process.info(:erlang.list_to_pid('<0.10137.4031>'))

[
  current_function: {:gen_server, :loop, 7},
  initial_call: {:proc_lib, :init_p, 5},
  status: :waiting,
  message_queue_len: 0,
  links: [],
  dictionary: [
    "$initial_call": {:brod_topic_subscriber, :init, 1},
    "$ancestors": [#PID<0.2211.0>, #PID<0.2208.0>, App.Supervisor,
     #PID<0.1959.0>]
  ],
  trap_exit: false,
  error_handler: :error_handler,
  priority: :normal,
  group_leader: #PID<0.1958.0>,
  total_heap_size: 10957,
  heap_size: 4185,
  stack_size: 10,
  reductions: 142808819,
  garbage_collection: [
    max_heap_size: %{error_logger: true, kill: true, size: 0},
    min_bin_vheap_size: 46422,
    min_heap_size: 233,
    fullsweep_after: 65535,
    minor_gcs: 78
  ],
  suspending: []
]

sys.get_state(:erlang.list_to_pid('<0.10137.4031>'))

{:state, :"kafka_consumer_client_some-group",
 #Reference<0.3846263541.660340737.50838>, "some-topic",
 [
   {:consumer, 3, #PID<0.2225.0>, #Reference<0.3846263541.660340737.50855>,
   28056, 28056}
 ], :brod_group_subscriber_worker,
 {:state,
  %{
    begin_offset: 26766,
    cb_config: nil,
    cb_module: App.Consumer,
    commit_fun: #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>,
    group_id: "some-group",
    partition: 3,
    topic: "some-topic"
  }, App.Consumer, nil,
  #Function<3.129769404/1 in :brod_group_subscriber_v2.maybe_start_worker/6>},
 :message_set}
```

28056 is highwater offset for this partition, but offsets are commited only to 27190

**Comment by epsylonix** [comment]
this last one topic subscriber is not connected to any `brod_group_subscriber_v2`
```
Process.info(:erlang.list_to_pid('<0.2211.0>'))    
nil
```

And because offsets are commited through `brod_group_subscriber_v2` using `gen_server:cast` but there is no one to cast to, offsets are not committed. 

So it seems that the previous `brod_group_subscriber_v2` started subscriber for this partition, then terminated but leaved the subscriber, the new subscriber couldn't subscribe, so all the messages are consumed by the old subscriber and since `brod_group_subscriber_v2` is not there anymore, those offsets are not committed. 

If I restart this instance now, all the messages will be processed again :unamused:

**Comment by k32** [comment]
Hi,

Which version of brod are you using? I recall fixing a similar problem in v3.13

**Comment by epsylonix** [comment]
sorry, forgot to tell up front, we're using brod v. 3.14.0

**Comment by k32** [comment]
Thanks for the info, I reproduced the error and found the root cause.

**Comment by epsylonix** [comment]
that was fast, thanks! Could you also publish this fix as a new version?

**Comment by k32** [comment]
Tl;dr: we created a community fork of brod and the relevant apps in a new org: https://github.com/kafka4beam/brod/
It may take a little while to set up the CI, though.

Just for the context: Kafka support for Erlang and Elixir had been developed by an internal team at Klarna. The team was dissolved long ago, and all core contributors have left the company since. Klarna github organization is only open to the employees, so the core designers were unable to interact with the code directly. We came into contact with them (@zmstone and @id) and we agreed that having a "community edition" of brod and the related apps makes sense.

**Comment by gugahoa** [comment]
Seems like it was fixed here https://github.com/kafka4beam/brod/commit/fff1136fbc6919b15d384a025c92eb11128d358d

**Comment by epsylonix** [comment]
Yes, it was fixed indeed. thank you.

**Comment by k32** [comment]
Hi everyone,
Sorry for slow responses, I will come back to brod development after the summer break.
